UnmanagedCallersOnly test failing in the outerloop during compilation

The issue seems to be due to some C# compiler update - the failures seem to be at places of negative testing that expects runtime failures, but we started to get compilation failures instead.

```
/__w/1/s/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs(341,6): error CS8896: 'UnmanagedCallersOnly' can only be applied to ordinary static methods or static local functions. [/__w/1/s/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.csproj]
##[error]src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs(341,6): error CS8896: 'UnmanagedCallersOnly' can only be applied to ordinary static methods or static local functions.
/__w/1/s/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs(439,6): error CS8895: Methods attributed with 'UnmanagedCallersOnly' cannot have generic type parameters and cannot be declared in a generic type. [/__w/1/s/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.csproj]
##[error]src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs(439,6): error CS8895: Methods attributed with 'UnmanagedCallersOnly' cannot have generic type parameters and cannot be declared in a generic type.
/__w/1/s/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs(440,48): error CS8894: Cannot use 'T' as a parameter type on a method attributed with 'UnmanagedCallersOnly'. [/__w/1/s/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.csproj]
##[error]src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs(440,48): error CS8894: Cannot use 'T' as a parameter type on a method attributed with 'UnmanagedCallersOnly'.
/__w/1/s/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs(520,10): error CS8895: Methods attributed with 'UnmanagedCallersOnly' cannot have generic type parameters and cannot be declared in a generic type. [/__w/1/s/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.csproj]
##[error]src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs(520,10): error CS8895: Methods attributed with 'UnmanagedCallersOnly' cannot have generic type parameters and cannot be declared in a generic type.
```

Log here:
https://dev.azure.com/dnceng/public/_build/results?buildId=811186&view=logs&j=db768a9e-3325-5ce3-2d52-896c40bee68a&t=59112ad9-fe58-56ee-6943-84df6f0e3f8e&l=13792

--- a/src/tests/Interop/UnmanagedCallersOnly/InvalidCSharp.ilproj
+++ b/src/tests/Interop/UnmanagedCallersOnly/InvalidCSharp.ilproj
@@ -1,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk.IL">
+  <PropertyGroup>
+    <OutputType>library</OutputType>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="InvalidCallbacks.il" />
+  </ItemGroup>
+</Project>

--- a/src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il
+++ b/src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il
@@ -4,7 +4,7 @@
 .assembly extern System.Runtime { }
 .assembly extern System.Runtime.InteropServices { }
 
-.assembly InvalidCSharp.dll { }
+.assembly InvalidCSharp { }
 
 .class public auto ansi beforefieldinit InvalidCSharp.GenericClass`1<T>
     extends System.Object

--- a/src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il
+++ b/src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il
@@ -1,0 +1,75 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+.assembly extern System.Runtime { }
+.assembly extern System.Runtime.InteropServices { }
+
+.assembly InvalidCSharp.dll { }
+
+.class public auto ansi beforefieldinit InvalidCSharp.GenericClass`1<T>
+    extends System.Object
+{
+    .method public hidebysig static 
+        void CallbackMethod (
+            int32 n
+        ) cil managed preservesig 
+    {
+        .custom instance void [System.Runtime.InteropServices]System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute::.ctor() = (
+            01 00 00 00
+        )
+        .maxstack 8
+        IL_0000: ldstr "Functions with attribute UnmanagedCallersOnlyAttribute within a generic type are invalid"
+        IL_0005: newobj instance void [System.Runtime]System.Exception::.ctor(string)
+        IL_000a: throw
+    }
+
+    .method public hidebysig specialname rtspecialname 
+        instance void .ctor () cil managed 
+    {
+        .maxstack 8
+        IL_0000: ldarg.0
+        IL_0001: call instance void [System.Runtime]System.Object::.ctor()
+        IL_0006: ret
+    }
+}
+
+.class public auto ansi beforefieldinit InvalidCSharp.Callbacks
+    extends [System.Runtime]System.Object
+{
+    .method public hidebysig static 
+        int32 CallbackMethodGeneric<T> (
+            !!T arg
+        ) cil managed preservesig 
+    {
+        .custom instance void [System.Runtime.InteropServices]System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute::.ctor() = (
+            01 00 00 00
+        )
+        .maxstack 8
+        IL_0000: ldstr "Functions with attribute UnmanagedCallersOnlyAttribute cannot have generic arguments"
+        IL_0005: newobj instance void [System.Runtime]System.Exception::.ctor(string)
+        IL_000a: throw
+    }
+
+    .method public hidebysig 
+        instance int32 CallbackNonStatic (
+            int32 val
+        ) cil managed preservesig 
+    {
+        .custom instance void [System.Runtime.InteropServices]System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute::.ctor() = (
+            01 00 00 00
+        )
+        .maxstack 8
+        IL_0000: ldstr "Instance functions with attribute UnmanagedCallersOnlyAttribute are invalid"
+        IL_0005: newobj instance void [System.Runtime]System.Exception::.ctor(string)
+        IL_000a: throw
+    }
+
+    .method public hidebysig specialname rtspecialname 
+        instance void .ctor () cil managed 
+    {
+        .maxstack 8
+        IL_0000: ldarg.0
+        IL_0001: call instance void [System.Runtime]System.Object::.ctor()
+        IL_0006: ret
+    }
+}

--- a/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs
+++ b/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs
@@ -3,9 +3,11 @@
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.Runtime.CompilerServices;
+using System.IO;
 using System.Reflection;
 using System.Reflection.Emit;
+using System.Runtime.CompilerServices;
+using System.Runtime.Loader;
 using System.Runtime.InteropServices;
 using System.Threading;
 using TestLibrary;
@@ -33,6 +35,20 @@ public class Program
         [UnmanagedCallersOnly]
         [DllImport(nameof(UnmanagedCallersOnlyDll), EntryPoint = "DoesntExist")]
         public static extern int PInvokeMarkedWithUnmanagedCallersOnly(int n);
+    }
+
+    private const string InvalidCSharpAssemblyName = "InvalidCSharp.dll";
+
+    public static Type GetCallbacksType()
+    {
+        var asm = AssemblyLoadContext.Default.LoadFromAssemblyPath(Path.Combine(Environment.CurrentDirectory, InvalidCSharpAssemblyName));
+        return asm.GetType("InvalidCSharp.Callbacks");
+    }
+
+    public static Type GetGenericClassOfIntType()
+    {
+        var asm = AssemblyLoadContext.Default.LoadFromAssemblyPath(Path.Combine(Environment.CurrentDirectory, InvalidCSharpAssemblyName));
+        return asm.GetType("InvalidCSharp.GenericClass`1").MakeGenericType(typeof(int));
     }
 
     private delegate int IntNativeMethodInvoker();
@@ -338,12 +354,6 @@ public class Program
         }
     }
 
-    [UnmanagedCallersOnly]
-    public int CallbackNonStatic(int val)
-    {
-        Assert.Fail($"Instance functions with attribute {nameof(UnmanagedCallersOnlyAttribute)} are invalid");
-        return -1;
-    }
 
     public static void NegativeTest_NonStaticMethod()
     {
@@ -354,7 +364,7 @@ public class Program
            {
                 .locals init ([0] native int ptr)
                 nop
-                ldftn      int CallbackNonStatic(int)
+                ldftn      int GetCallbacksType().CallbackNonStatic(int)
                 stloc.0
 
                 ldloc.0
@@ -371,7 +381,7 @@ public class Program
         il.Emit(OpCodes.Nop);
 
         // Get native function pointer of the callback
-        il.Emit(OpCodes.Ldftn, typeof(Program).GetMethod(nameof(CallbackNonStatic)));
+        il.Emit(OpCodes.Ldftn, GetCallbacksType().GetMethod("CallbackNonStatic"));
         il.Emit(OpCodes.Stloc_0);
         il.Emit(OpCodes.Ldloc_0);
 
@@ -436,13 +446,6 @@ public class Program
         Assert.Throws<InvalidProgramException>(() => { testNativeMethod(); });
     }
 
-    [UnmanagedCallersOnly]
-    public static int CallbackMethodGeneric<T>(T arg)
-    {
-        Assert.Fail($"Functions with attribute {nameof(UnmanagedCallersOnlyAttribute)} cannot have generic arguments");
-        return -1;
-    }
-
     public static void NegativeTest_NonInstantiatedGenericArguments()
     {
         Console.WriteLine($"Running {nameof(NegativeTest_NonInstantiatedGenericArguments)}...");
@@ -452,7 +455,7 @@ public class Program
            {
                 .locals init ([0] native int ptr)
                 IL_0000:  nop
-                IL_0001:  ldftn      void CallbackMethodGeneric(T)
+                IL_0001:  ldftn      void InvalidCSharp.Callbacks.CallbackMethodGeneric(T)
                 IL_0007:  stloc.0
                 IL_0008:  ret
              }
@@ -463,7 +466,7 @@ public class Program
         il.Emit(OpCodes.Nop);
 
         // Get native function pointer of the callback
-        il.Emit(OpCodes.Ldftn, typeof(Program).GetMethod(nameof(CallbackMethodGeneric)));
+        il.Emit(OpCodes.Ldftn, GetCallbacksType().GetMethod("CallbackMethodGeneric"));
         il.Emit(OpCodes.Stloc_0);
 
         il.Emit(OpCodes.Ret);
@@ -482,7 +485,7 @@ public class Program
            {
                 .locals init ([0] native int ptr)
                 nop
-                ldftn      void CallbackMethodGeneric(int)
+                ldftn      void InvalidCSharp.Callbacks.CallbackMethodGeneric(int)
                 stloc.0
 
                 ldloc.0
@@ -499,7 +502,7 @@ public class Program
         il.Emit(OpCodes.Nop);
 
         // Get native function pointer of the instantiated generic callback
-        il.Emit(OpCodes.Ldftn, typeof(Program).GetMethod(nameof(CallbackMethodGeneric)).MakeGenericMethod(new [] { typeof(int) }));
+        il.Emit(OpCodes.Ldftn, GetCallbacksType().GetMethod("CallbackMethodGeneric").MakeGenericMethod(new [] { typeof(int) }));
         il.Emit(OpCodes.Stloc_0);
         il.Emit(OpCodes.Ldloc_0);
 
@@ -515,15 +518,6 @@ public class Program
         Assert.Throws<InvalidProgramException>(() => { testNativeMethod(); });
     }
 
-    public class GenericClass<T>
-    {
-        [UnmanagedCallersOnly]
-        public static void CallbackMethod(int n)
-        {
-            Assert.Fail($"Functions with attribute {nameof(UnmanagedCallersOnlyAttribute)} within a generic type are invalid");
-        }
-    }
-
     public static void NegativeTest_FromInstantiatedGenericClass()
     {
         Console.WriteLine($"Running {nameof(NegativeTest_FromInstantiatedGenericClass)}...");
@@ -533,7 +527,7 @@ public class Program
            {
                 .locals init ([0] native int ptr)
                 nop
-                ldftn      int GenericClass<int>::CallbackMethod(int)
+                ldftn      int InvalidCSharp.GenericClass<int>::CallbackMethod(int)
                 stloc.0
 
                 ldloc.0
@@ -550,7 +544,7 @@ public class Program
         il.Emit(OpCodes.Nop);
 
         // Get native function pointer of the callback from the instantiated generic class.
-        il.Emit(OpCodes.Ldftn, typeof(GenericClass<int>).GetMethod(nameof(GenericClass<int>.CallbackMethod)));
+        il.Emit(OpCodes.Ldftn, GetGenericClassOfIntType().GetMethod("CallbackMethod"));
         il.Emit(OpCodes.Stloc_0);
         il.Emit(OpCodes.Ldloc_0);
 

--- a/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs
+++ b/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs
@@ -35,17 +35,17 @@ public class Program
         public static extern int PInvokeMarkedWithUnmanagedCallersOnly(int n);
     }
 
-    private const string InvalidCSharpAssemblyName = "InvalidCSharp.dll";
+    private const string InvalidCSharpAssemblyName = "InvalidCSharp";
 
     public static Type GetCallbacksType()
     {
-        var asm = Assembly.Load("InvalidCSharp");
+        var asm = Assembly.Load(InvalidCSharpAssemblyName);
         return asm.GetType("InvalidCSharp.Callbacks");
     }
 
     public static Type GetGenericClassOfIntType()
     {
-        var asm = Assembly.Load("InvalidCSharp");
+        var asm = Assembly.Load(InvalidCSharpAssemblyName);
         return asm.GetType("InvalidCSharp.GenericClass`1").MakeGenericType(typeof(int));
     }
 

--- a/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs
+++ b/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs
@@ -3,11 +3,9 @@
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.IO;
 using System.Reflection;
 using System.Reflection.Emit;
 using System.Runtime.CompilerServices;
-using System.Runtime.Loader;
 using System.Runtime.InteropServices;
 using System.Threading;
 using TestLibrary;
@@ -41,13 +39,13 @@ public class Program
 
     public static Type GetCallbacksType()
     {
-        var asm = AssemblyLoadContext.Default.LoadFromAssemblyPath(Path.Combine(Environment.CurrentDirectory, InvalidCSharpAssemblyName));
+        var asm = Assembly.Load("InvalidCSharp");
         return asm.GetType("InvalidCSharp.Callbacks");
     }
 
     public static Type GetGenericClassOfIntType()
     {
-        var asm = AssemblyLoadContext.Default.LoadFromAssemblyPath(Path.Combine(Environment.CurrentDirectory, InvalidCSharpAssemblyName));
+        var asm = Assembly.Load("InvalidCSharp");
         return asm.GetType("InvalidCSharp.GenericClass`1").MakeGenericType(typeof(int));
     }
 

--- a/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.csproj
+++ b/src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.csproj
@@ -1,7 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <CLRTestPriority>1</CLRTestPriority>
   </PropertyGroup>
   <ItemGroup>
     <Compile Include="UnmanagedCallersOnlyTest.cs" />
@@ -10,5 +9,6 @@
     <!-- This is needed to make sure native binary gets installed in the right location -->
     <ProjectReference Include="CMakeLists.txt" />
     <ProjectReference Include="$(TestSourceDir)Common/CoreCLRTestLibrary/CoreCLRTestLibrary.csproj" />
+    <ProjectReference Include="InvalidCSharp.ilproj" />
   </ItemGroup>
 </Project>

